### ISSUE TITLE
Will Remix-PWA support the move to React Router v7?

### ISSUE THREAD

**Issue by henryStelle** [remix-pwa/monorepo#284]
Hi,
This seems like a great package. However, it seems Remix is "taking a pause" now that all the functionality has moved to React Router v7:
- https://remix.run/blog/merging-remix-and-react-router
- https://remix.run/blog/react-router-v7

Will Remix-PWA publish future versions based around React Router v7 instead of Remix?

Thanks!

**Comment by ShafSpecs** [comment]
Remix PWA would support react router as well as Remix. React Router v7 is very similar to Remix, and my goal is to ensure wether you choose to use existing Remix versions, or new React router versions, Remix PWA would provide support for them

**Comment by sindras** [comment]
Hi, thanks for the hard work here! 

Just curious if there is a timeline in mind?

I tried remix-pwa v3 today and noticed that it was not compatible with react-router v7, because remix-pwa made assumption that `remix-run/react` would be installed in the project. In addition, the documentation website has not been up-to-date too: it mentions v3, and v4, but the github repo itself is in the middle of releasing v5?

As someone new to do this package, I would love to know whether v5 is almost ready, or should I give v4 a try?

**Comment by ShafSpecs** [comment]
The docs also has a next section that's dedicated to v5, the docs aren't fully done yet tho. Regarding the timeline, I don't have one. React Router v7 is similar to Remix, but with a few distinctions that I need to explore

**Comment by IgnisDa** [comment]
Hey @ShafSpecs. Any movement on this?

**Comment by ShafSpecs** [comment]
> Hey @ShafSpecs. Any movement on this?

Yeah, it's been pretty slow and would be a dev release for a while, but I am working on this

**Comment by IgnisDa** [comment]
@ShafSpecs Is there a github branch i can try out for this?

**Comment by ShafSpecs** [comment]
> [@ShafSpecs](https://github.com/ShafSpecs) Is there a github branch i can try out for this?

Not yet unfortunately, I am still trying to get React Router v7 to work on my machine

**Comment by IgnisDa** [comment]
Maybe I can help? Can you post the error? I have it working on a new react-router app, as well as upgraded a big remix application to it.

**Comment by ShafSpecs** [comment]
> Maybe I can help? Can you post the error? I have it working on a new react-router app, as well as upgraded a big remix application to it.

This branch (v5-rc): https://github.com/remix-pwa/monorepo/tree/v5-rc/sandboxes/rr7

I re-worked the sandbox to contain remix spa apps, normal ssr apps and a react router v7 app. The Remix apps work, but React Router throws errors for some reason

```
Unexpected Server Error

Error: Objects are not valid as a React child (found: object with keys {$$typeof, type, key, ref, props, _owner, _store}). If you meant to render a collection of children, use an array instead.
```

**Comment by IgnisDa** [comment]
Okay i got it working. Seems to be a problem with the sandbox being inside a workspace. Changes i made:

- root package.json 

<img width="980" alt="Image" src="https://github.com/user-attachments/assets/8ddf65cd-6954-4f6c-bc65-5c7ba615c26f" />

- run `touch package-lock.json && npm i` in `sandboxes/rr7`

- in `sandboxes/rr7/app/root.tsx`

<img width="979" alt="Image" src="https://github.com/user-attachments/assets/6885c6a5-c1c9-4b67-87a6-b076b3382d2f" />

I suspect that this problem exists only for npm workspaces since https://github.com/remix-run/react-router-templates uses pnpm which works. and I can tell from personal experience that yarn workspaces also work.

Hope that points you in the right direction...

**Comment by ShafSpecs** [comment]
Thanks for the fix @IgnisDa! Started work on `reactRouterPWA` plugin

**Comment by IgnisDa** [comment]
Hi @ShafSpecs! Any updates on this issue?

**Comment by ShafSpecs** [comment]
Yeah, I'm still working on the manor PR for react router.
https://github.com/remix-pwa/monorepo/pull/287

This update would support both Remix and React Router, and then a second release would be coming after to sunset Remix support. With project renaming and a few other bonuses

**Comment by IgnisDa** [comment]
Great to know. Thanks for the update!

**Comment by sjbuysse** [comment]
I know it's not easy to do, but do you have any ballpark estimates on release date?

**Comment by ShafSpecs** [comment]
> I know it's not easy to do, but do you have any ballpark estimates on release date? 

I'm currently pretty slow :sweat_smile:, but once I clear one of my freelance works (planning to wrap that up this month), I can slot this full time into my schedule

**Comment by iwan-uschka** [comment]
If there is something i (or we) can help with, let me (us) know.

**Comment by ShafSpecs** [comment]
> If there is something i (or we) can help with, let me (us) know.

Thanks for the offer! Currently, one of the last things remaining is lots of testing. I can push to dev later today, so it's available and you can start using (and testing real-time) on your end 👌. Most of the apis are the same, just a few changes

**Comment by ShafSpecs** [comment]
Merged the first major change, I would be pushing more fixes and features in time

**Comment by ShafSpecs** [comment]
Before I go update the docs, I would push the current changes to `latest`.

Create a new release for react-router-pwa, remove Remix support and then go revamp the docs. Sounds good?

**Comment by IgnisDa** [comment]
Yep that sounds like a good plan. Maybe also a canary release on npm so
that users can try it out and help you iron out the rough edges?

**Comment by ShafSpecs** [comment]
The canary release is out (via the `dev` tag). That has been published already

**Comment by jack-r-warren** [comment]
This is awesome work, truly -- the most immediate thing coming up in testing for me is `@remix-pwa/worker-runtime@3.0.0-dev.1` referencing `@remix-run/server-runtime` (not in my RR7 project) for its old `json(..)` utility.

```
vite v6.1.1 building for production...
🏗️  Building Service Worker in production mode...
x Build failed in 136ms
Error during worker build: ../node_modules/@remix-pwa/worker-runtime/dist/src/utils/response.js (1:9): "json" is not exported by "__vite-optional-peer-dep:@remix-run/server-runtime/dist/responses.js:@remix-pwa/worker-runtime", imported by "../node_modules/@remix-pwa/worker-runtime/dist/src/utils/response.js".
```

`@remix-pwa/sw@4.0.0-dev.2` seems like it has a heavier use of Remix dependencies so it's 100% possible this is user error on my part.

**Comment by ShafSpecs** [comment]
I would need to fix that ASAP. Thx for the heads up

**Comment by IgnisDa** [comment]
To add, I also get this error while building the project: 

```
Error during worker build: [vite]: Rollup failed to resolve import "@remix-run/react" from "/Users/diptesh/Desktop/Code/ryot/apps/frontend/node_modules/@remix-pwa/sw/dist/src/utils/installPWAGlobals.js".
```

This is probably due to the usage of `@remix-run/react` instead of `react-router` in `packages/dev/react-router/*`.

**Comment by alexfoxy** [comment]
Look forward to trying this, we want to make the switch.

**Comment by ShafSpecs** [comment]
Opened a new PR, still making sure I don't nuke anyone's project. This PR fully sunsets Remix support fully (as dual support had so many issues)

https://github.com/remix-pwa/monorepo/pull/292

**Comment by ShafSpecs** [comment]
React Router PWA v5 vs React Router PWA v1?

**Comment by IgnisDa** [comment]
Do you plan to rename the project to react-router-pwa? I see the package is free on npm (https://www.npmjs.com/package/react-router-pwa).

If you do plan to rename, i recommend pushing a few dummy files to that package so that it is not claimed.

**Comment by IgnisDa** [comment]
> React Router PWA v5 vs React Router PWA v1?

I'd personally prefer v1 since the new project targets react-router and not remix. so its a new project IMO.

**Comment by ShafSpecs** [comment]
> Do you plan to rename the project to react-router-pwa? I see the package is free on npm (https://www.npmjs.com/package/react-router-pwa).

Yes, keeping the name Remix PWA becomes confusing in the long run. I'm also debating renaming the org or just creating another and archiving the code here. The first is the easiest, the second allows for legacy support

**Comment by sjbuysse** [comment]
any update on this? eager to give it a try in my rr7 project

**Comment by iwan-uschka** [comment]
@sjbuysse, please see @ShafSpecs's https://github.com/remix-pwa/monorepo/issues/284#issuecomment-2692658099:

> Currently, one of the last things remaining is lots of testing.

So yes, give it a try and provide feedback if you can.

**Comment by sjbuysse** [comment]
> [@sjbuysse](https://github.com/sjbuysse), please see [@ShafSpecs](https://github.com/ShafSpecs)'s [#284 (comment)](https://github.com/remix-pwa/monorepo/issues/284#issuecomment-2692658099):
> 
> > Currently, one of the last things remaining is lots of testing.
> 
> So yes, give it a try and provide feedback if you can.

Pardon my ignorance, but how do I install the dev version?

**Comment by IgnisDa** [comment]
> > Do you plan to rename the project to react-router-pwa? I see the package is free on npm (https://www.npmjs.com/package/react-router-pwa).
> 
> Yes, keeping the name Remix PWA becomes confusing in the long run. I'm also debating renaming the org or just creating another and archiving the code here. The first is the easiest, the second allows for legacy support

@ShafSpecs were you able to arrive on a decision for this?

**Comment by ShafSpecs** [comment]
Yep, I would try and lock down the package today. Irl is super busy at the moment. Currently getting ready to head out to the airport lol

So please, forgive my tardiness

**Comment by IgnisDa** [comment]
Oh no, please always prioritize your life first! Thanks for the update.

**Comment by danvoyce** [comment]
Hey just wondering if there were any updates / ETA when this will be resolved?

Anything we can do to help?

**Comment by SamuMazzi** [comment]
Hi @ShafSpecs, I tried the package and I think it all worked well pretty fine except for this error that I had
```
🏗️  Building Service Worker in production mode...
✗ Build failed in 491ms
Error during worker build: node_modules/@remix-pwa/worker-runtime/dist/src/utils/response.js (1:9): "json" is not exported by "__vite-optional-peer-dep:@remix-run/server-runtime/dist/responses.js:@remix-pwa/worker-runtime", imported by "node_modules/@remix-pwa/worker-runtime/dist/src/utils/response.js".
```

But I fixed it simply modifying `@remix-pwa/worker-runtime/dist/src/utils/response.js` this way
```diff
- import { json } from '@remix-run/server-runtime/dist/responses.js';

export function errorResponseToJson(errorResponse) {
    // @ts-expect-error
-   return json(errorResponse.error || { message: 'Unexpected Server Error' }, {
+   return errorResponse.error || { message: 'Unexpected Server Error' }, {
        status: errorResponse.status,
        statusText: errorResponse.statusText,
        headers: {
            'X-Remix-Error': 'yes',
        },
-   });
+   };
}
```

I don't know if it's the right way to do it or if this is the right place to write this, just trying to be helpful :)

**Comment by ShafSpecs** [comment]
> Hi @ShafSpecs, I tried the package and I think it all worked well pretty fine except for this error that I had
> ```
> 🏗️  Building Service Worker in production mode...
> ✗ Build failed in 491ms
> Error during worker build: node_modules/@remix-pwa/worker-runtime/dist/src/utils/response.js (1:9): "json" is not exported by "__vite-optional-peer-dep:@remix-run/server-runtime/dist/responses.js:@remix-pwa/worker-runtime", imported by "node_modules/@remix-pwa/worker-runtime/dist/src/utils/response.js".
> ```
> 
> But I fixed it simply modifying `@remix-pwa/worker-runtime/dist/src/utils/response.js` this way
> ```diff
> - import { json } from '@remix-run/server-runtime/dist/responses.js';
> 
> export function errorResponseToJson(errorResponse) {
>     // @ts-expect-error
> -   return json(errorResponse.error || { message: 'Unexpected Server Error' }, {
> +   return errorResponse.error || { message: 'Unexpected Server Error' }, {
>         status: errorResponse.status,
>         statusText: errorResponse.statusText,
>         headers: {
>             'X-Remix-Error': 'yes',
>         },
> -   });
> +   };
> }
> ```
> 
> I don't know if it's the right way to do it or if this is the right place to write this, just trying to be helpful :)

Thx, you can open a PR to the dev branch with the fix. Thank you. To push as soon as a merge occurs, use the commit message format: "fix(worker-runtime): custom message". Thank you ❤️

**Comment by Moonflower-labs** [comment]
Hi, hope you’re well. Any idea on when thIs will be merged and ready to use with react router v7? Thanks 🙏🏼

**Comment by zakinadhif** [comment]
waiting for react router v7 support too ♥️

**Comment by ShafSpecs** [comment]
> Hi, hope you’re well. Any idea on when thIs will be merged and ready to use with react router v7? Thanks 🙏🏼 

Hello 👋. I can't say :(. I hardly work with web technologies nowadays, so my projects are not tailored towards React/Web ecosystem. I know Sanchez is currently working on supporting RRv7 in Vite PWA

That being said, I plan to give this project one last major update, but I can't put a date on it

**Comment by iwan-uschka** [comment]
For the record: 👆https://github.com/vite-pwa/vite-plugin-pwa/issues/809

**Comment by userquin** [comment]
I don't use React, rn too busy at work, here a draft pr that is working with rr7, the repo should be transferred to vite pwa org. once ready (still a lot of work there):

https://github.com/userquin/vite-pwa-react-router/pull/3

**Comment by IgnisDa** [comment]
@userquin Thanks for this! Was there any movement on this re-work?
